File: 0352-data-stream-as-disjoint-intervals/0352-data-stream-as-disjoint-intervals.java
class SummaryRanges {
    private TreeMap<Integer, int[]> map = new TreeMap<>();
    public SummaryRanges() {
        
    }
    
    public void addNum(int val) {
        if (map.containsKey(val))
          return;

        final Integer lo = map.lowerKey(val);  
        final Integer hi = map.higherKey(val); 

       
        if (lo != null && hi != null && map.get(lo)[1] + 1 == val && val + 1 == hi) {
          map.get(lo)[1] = map.get(hi)[1];
          map.remove(hi);
        } else if (lo != null && map.get(lo)[1] + 1 >= val) {
          map.get(lo)[1] = Math.max(map.get(lo)[1], val);
        } else if (hi != null && val + 1 == hi) {
          map.put(val, new int[] {val, map.get(hi)[1]});
          map.remove(hi);
        } else {
          map.put(val, new int[] {val, val});
        }
    }
    public int[][] getIntervals() {
        List<int[]> intervals = new ArrayList<>(map.values());
        return intervals.toArray(new int[intervals.size()][]);
    }
}

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(value);
 * int[][] param_2 = obj.getIntervals();
 */

File: 0352-data-stream-as-disjoint-intervals/0352-data-stream-as-disjoint-intervals.cpp
class SummaryRanges {
private:
    map<int,int>mp;
public:
    SummaryRanges() {
        
    }
    
    void addNum(int value) {
        auto it = mp.lower_bound(value);
        bool merged = false;
        if(it != mp.begin()) {
            auto prev = it;
            --prev;
            if(prev->second + 1 >= value) {
                merged = true;
                prev->second = max(prev->second, value);
            }
        }

        if(it != mp.end()) {
            if(it->first - 1 <= value) {
                if(merged) {
                    auto prev = it;
                    --prev;
                    if(prev->second >= it->first - 1) {
                        prev->second = max(prev->second, it->second);
                        mp.erase(it);
                    }
                } else {
                    merged = true;
                    if(it->first != value) {
                        pair<int, int> p = *it;
                        p.first = min(p.first, value);
                        it = mp.insert(it, p);
                        ++it;
                        if(it != mp.end())
                            mp.erase(it);
                    }
                }
            }
        }
        if(!merged)
            mp.insert(it, {value, value});
    }
    vector<vector<int>> getIntervals() {
        vector<vector<int>> intervals;
        for(auto const & p : mp)
            intervals.push_back({p.first, p.second});
        return intervals;
    }
};


/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges* obj = new SummaryRanges();
 * obj->addNum(value);
 * vector<vector<int>> param_2 = obj->getIntervals();
 */